### PR TITLE
Support cli

### LINKED ISSUE #13: RFF, Download + ls
## Context
Can we add a ls command?
Download should not return a json, it should return the file or encode the response in bas64 or similar

@eordano 
What would be different between a `ls` vs current `dependencies/ipfs` (in branch `development`)??

## Features
- endpoint for download
- endpoint for ls TBD

### PR DESCRIPTION
**Pull request by nachomazzara**
# Context

- Endpoints to support CLI and Client
- Check request within the lens' blacklist
- Check land metadata directly on contract
- Cache request responses and simulate a DHT on REDIS
- Error handlers & responses

# Issues / Tickets

Fix #13 

# Notes

Replace this [PR](https://github.com/decentraland/ipfs-node/pull/12)

### CODE REVIEW

File: src/ethereum.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  try {
+    await web3Eth.disconnect() // clean if it is a retry
+    let connected = await web3Eth.connect({
+      contracts: [LANDRegistry]
+    })
+    if (!connected) {
+      throw new Error('Could not connect to the blockchain')
+    }
+  } catch (e) {
+    if (e.message.indexOf('Could not connect to the blockchain') !== -1) {
```
Comment (andresmijares) [inline src/ethereum.js]: if this statement necessary? what would be another case where we wouldnt require to try to re-connect?

File: src/ethereum.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  try {
+    const land = web3Eth.getContract('LANDRegistry')
+    const metadata = await land.landData(x, y)
+    ipns = await LANDRegistry.decodeLandData(metadata).ipns
+    return ipns ? ipns.split(':')[1] : 0
+  } catch (e) {
+    // return undefined ipns
+    console.log(e.message)
+  }
+  return ipns
```
Comment (andresmijares) [inline src/ethereum.js]: ?? this returning `undefined` is un clear, that `let ipns` looks weird outside the try/catch, can lead to code smell, is that necessary? if you really need it, you can a finally statement and return `undefined` or `null` prefer.

File: src/ipfs.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      try {
+        const ipfs = req.params.ipfs
+        const file = req.params[0] ? `${ipfs}/${req.params[0]}` : ipfs
+
+        const blackListed = await isHashBlacklisted(ipfs)
+        if (blackListed) {
+          throw new Error(`IPFS ${ipfs} is blacklisted`)
+        }
+
+        request.get(`http://localhost:8080/ipfs/${file}`).pipe(res)
```
Comment (andresmijares) [inline src/ipfs.js]: hard coded url

File: src/ipfs.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        if (!ipns) {
+          throw new Error('The land has not IPNS associated')
+        }
+        const ipfs = await Download.resolveIPNS(ipns)
+        const dependencies = await Download.resolveDependencies(ipfs)
+        const url = { ipns, ipfs, dependencies }
+        await setParcel({ x, y }, url)
+        return res.json({ ok: true, url })
+      } catch (error) {
+        return res.json({ ok: false, error: error.message })
```
Comment (andresmijares) [inline src/ipfs.js]: does it add a status code `403`? I cannot recall this.

Comment (nachomazzara) [reply src/ipfs.js]: Yeah, lets add an errorHandling too

File: src/ipfs.js (hunk 3 above)
Comment (nachomazzara) [reply src/ipfs.js]: 👌

File: src/ethereum.js (hunk 1 above)
Comment (nachomazzara) [reply src/ethereum.js]: Nope, it was a copy & paste from another project

File: src/ethereum.js (hunk 2 above)
Comment (nachomazzara) [reply src/ethereum.js]: We could return `null` at `catch` clause.

File: src/blacklist.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+const axios = require('axios')
+const createError = require('http-errors')
+
+async function checkIPFS (ipfs) {
+  if (process.env.BLACKLIST_URL) {
+    const res = await axios.get(`${process.env.BLACKLIST_URL}/blacklist/${ipfs}`)
+    const { blacklisted } = res.data
+    if (blacklisted) {
+      throw createError(403, `IPFS ${ipfs} is blacklisted`)
```
Comment (andresmijares) [inline src/blacklist.js:9]: me gusto esto

**Review by andresmijares — APPROVED** [review]
excelente

**Pull request merged**